### ISSUE TITLE
does small batch size can affect accuracy？

### ISSUE THREAD

**Issue by verigle** [davidnvq/grit#15]
I only can  use batch size = 4 to run train_caption because I only have one GPU

**Comment by davidnvq** [comment]
I haven't trained on 1 GPU. Of course it will not reproduce the results because the learning rate is optimal for `batch_size = 8*16`. Thus, you can either:
1. keep the same batch size = 4 but using [a simple trick](https://medium.com/huggingface/training-larger-batches-practical-tips-on-1-gpu-multi-gpu-distributed-setups-ec88c3e51255) like `accumulation_steps` in `train_xe` and `train_sc` in [caption_engine.py](https://github.com/davidnvq/grit/blob/550cae43cd109315e2d2701dff2187752f031dfd/engine/caption_engine.py).
2. or you have to find the optimal learning rate for small batch size. The rule for scaling the learning rate using SGD is simple while it does not hold true for Adam. Thus, I suggest you implement the first solution.

**Comment by davidnvq** [comment]
I am currently able to release the training log with 1 GPU for the entire dataset using this code (up to 13 epochs). **It runs properly, performance seems acceptable and keeps increasing.** However, I think hyperparameters are not optimal for small batch size. It may need the trick of **accumulating gradients**. The below result is obtained with freezed detector, batch size 8 only.

```
exp          backbone   imsize    resize   raug   epoch   split   cider    B1      B4      R       M       B2      B3      t-loss   t-reward   b-reward   which   v-loss
freeze_all   B-VG       384_640   maxwh    True   0       valid   103.42   73.92   33.00   54.45   25.92   57.83   43.80   3.22     0.00       0.00       fr_xe   2.55
freeze_all   B-VG       384_640   maxwh    True   0       test    103.65   73.65   32.66   54.32   25.80   57.44   43.44   3.22     0.00       0.00       fr_xe   2.55
freeze_all   B-VG       384_640   maxwh    True   1       valid   110.98   75.69   34.97   56.14   27.33   59.54   45.66   2.48     0.00       0.00       fr_xe   2.34
freeze_all   B-VG       384_640   maxwh    True   1       test    111.42   75.46   34.36   55.73   27.05   59.05   45.07   2.48     0.00       0.00       fr_xe   2.34
freeze_all   B-VG       384_640   maxwh    True   2       valid   114.31   76.68   35.89   56.95   27.70   60.84   46.82   2.33     0.00       0.00       fr_xe   2.28
freeze_all   B-VG       384_640   maxwh    True   2       test    115.03   76.69   35.86   56.70   27.54   60.63   46.72   2.33     0.00       0.00       fr_xe   2.28
freeze_all   B-VG       384_640   maxwh    True   3       valid   114.20   76.26   35.60   56.56   27.60   60.18   46.29   2.26     0.00       0.00       fr_xe   2.24
freeze_all   B-VG       384_640   maxwh    True   3       test    114.73   76.25   35.21   56.28   27.54   60.02   45.99   2.26     0.00       0.00       fr_xe   2.24
freeze_all   B-VG       384_640   maxwh    True   4       valid   117.32   77.59   36.79   57.31   27.99   61.81   47.80   2.20     0.00       0.00       fr_xe   2.23
freeze_all   B-VG       384_640   maxwh    True   4       test    117.73   77.41   36.76   57.01   27.82   61.64   47.78   2.20     0.00       0.00       fr_xe   2.23
freeze_all   B-VG       384_640   maxwh    True   5       valid   117.43   77.49   36.77   57.24   28.20   61.64   47.71   2.16     0.00       0.00       fr_xe   2.21
freeze_all   B-VG       384_640   maxwh    True   5       test    118.11   77.09   36.63   57.14   27.95   61.27   47.50   2.16     0.00       0.00       fr_xe   2.21
freeze_all   B-VG       384_640   maxwh    True   6       valid   115.45   76.81   35.97   56.86   27.98   60.42   46.56   2.13     0.00       0.00       fr_xe   2.19
freeze_all   B-VG       384_640   maxwh    True   6       test    117.25   76.84   36.31   56.96   28.04   60.72   46.97   2.13     0.00       0.00       fr_xe   2.19
freeze_all   B-VG       384_640   maxwh    True   7       valid   118.88   77.54   37.07   57.39   28.18   61.76   47.91   2.11     0.00       0.00       fr_xe   2.21
freeze_all   B-VG       384_640   maxwh    True   7       test    119.77   77.48   37.07   57.38   28.19   61.69   47.84   2.11     0.00       0.00       fr_xe   2.21
freeze_all   B-VG       384_640   maxwh    True   8       valid   117.11   77.07   36.95   57.09   28.17   61.34   47.62   2.09     0.00       0.00       fr_xe   2.20
freeze_all   B-VG       384_640   maxwh    True   8       test    117.30   76.64   36.31   56.86   27.98   60.83   47.05   2.09     0.00       0.00       fr_xe   2.20
freeze_all   B-VG       384_640   maxwh    True   9       valid   118.32   77.47   37.18   57.35   28.17   61.74   47.92   2.06     0.00       0.00       fr_xe   2.20
freeze_all   B-VG       384_640   maxwh    True   9       test    120.29   77.79   37.27   57.52   28.32   61.95   48.06   2.06     0.00       0.00       fr_xe   2.20
freeze_all   B-VG       384_640   maxwh    True   10      valid   129.65   81.61   40.15   59.21   28.90   66.77   52.31   -0.01    1.24       1.24       fr_sc   2.72
freeze_all   B-VG       384_640   maxwh    True   10      test    130.61   81.68   40.26   59.17   28.95   66.71   52.34   -0.01    1.24       1.24       fr_sc   2.72
freeze_all   B-VG       384_640   maxwh    True   11      valid   130.18   81.85   40.29   59.18   29.04   66.87   52.44   -0.02    1.27       1.27       fr_sc   3.55
freeze_all   B-VG       384_640   maxwh    True   11      test    132.26   82.19   40.34   59.26   29.05   67.03   52.55   -0.02    1.27       1.27       fr_sc   3.55
```

**Comment by davidnvq** [comment]
FYI, when training with extracted features stored in the HDF5 file using **1 GPU**, I notify the serious speed problem of HDF5 format.
When training with 5,000 valid images saved in the HDF5 file (for debugging), it **takes only 1.5 minutes**.
We expect that when training on the entire training images (~100k images) save in the HDF5 file , it will take about **20 x  1.5** minutes = **30** minutes.
Reality reveals that it is FALSE. It takes **7h 25m for 1 epoch.**  It is terrible.
> When I use multiple GPUs, reading HDF5 becomes much faster. 1 epoch only takes 2-5 minutes with 8 GPUs. The reason may be that the entire dataset is divided into smaller subsets, say /8 when training with 8 GPUs in DPP.

![image](https://user-images.githubusercontent.com/30741785/190049654-26cc7d6b-14dc-4128-8295-57fda010cbb9.png)

When I do more research about the issue of HDF5, I figured out that random access from the HDF5 becomes much slower when the size of file increases or the chunk size we want to retrieve. Googling will give us some cases where random accessing will slow down 20 times compared with ordered access.

A simple trick to gain much faster retrieval is that DO NOT use `shuffle` in dataloaders. However, it is not a good practice when training neural networks. It is recommended other formats of extracted features, e.g., 1 npy file for every image when you only have a single GPU?

Any insight here is welcome.
